tests/integration/ha_tests/helpers.py

```diff
@@ -0,0 +1,337 @@
+            action = await asyncio.wait_for(action.wait(), 10)
+            return action.results["primary"]
+
+
+async def is_replica(ops_test: OpsTest, unit_name: str) -> bool:
+    """Returns whether the unit a replica in the cluster."""
+    unit_ip = await get_unit_address(ops_test, unit_name)
+    member_name = unit_name.replace("/", "-")
+
+    try:
+        for attempt in Retrying(stop=stop_after_delay(60 * 3), wait=wait_fixed(3)):
+            with attempt:
+                cluster_info = requests.get(f"http://{unit_ip}:8008/cluster")
+
+                # The unit may take some time to be listed on Patroni REST API cluster endpoint.
+                if member_name not in {
+                    member["name"] for member in cluster_info.json()["members"]
+                }:
+                    raise MemberNotListedOnClusterError()
+
+                for member in cluster_info.json()["members"]:
+                    if member["name"] == member_name:
+                        role = member["role"]
+
+                # A member that restarted has the DB process stopped may
+                # take some time to know that a new primary was elected.
+                if role == "replica":
+                    return True
+                else:
+                    raise MemberNotUpdatedOnClusterError()
+    except RetryError:
+        return False
+
+
+async def postgresql_ready(ops_test, unit_name: str) -> bool:
+    """Verifies a PostgreSQL instance is running and available."""
+    unit_ip = await get_unit_address(ops_test, unit_name)
+    try:
+        for attempt in Retrying(stop=stop_after_delay(60 * 5), wait=wait_fixed(3)):
+            with attempt:
+                instance_health_info = requests.get(f"http://{unit_ip}:8008/health")
+                assert instance_health_info.status_code == 200
+    except RetryError:
+        return False
+
+    return True
+
+
+async def secondary_up_to_date(ops_test: OpsTest, unit_name: str, expected_writes: int) -> bool:
+    """Checks if secondary is up-to-date with the cluster.
+
+    Retries over the period of one minute to give secondary adequate time to copy over data.
+    """
+    app = await app_name(ops_test)
+    password = await get_password(ops_test, app)
+    status = await ops_test.model.get_status()
+    host = status["applications"][app]["units"][unit_name]["address"]
+    connection_string = (
+        f"dbname='application' user='operator'"
+        f" host='{host}' password='{password}' connect_timeout=10"
+    )
+
+    try:
+        for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(3)):
+            with attempt:
+                with psycopg2.connect(
+                    connection_string
+                ) as connection, connection.cursor() as cursor:
+                    cursor.execute("SELECT COUNT(number) FROM continuous_writes;")
+                    secondary_writes = cursor.fetchone()[0]
+                    assert secondary_writes == expected_writes
```

Comment:
is this just checking the last write that was written to the database or all writes? I don't know SQL well enough 😅

Reply:
This gets the total amount of numbers written (to compare with the last write that was returned after stopping the writes from the continuous writes charm).